File: ruby/consecutive_substrings.rb
def consecutive_substrings(string)
  # type your code in here
  subs = []

  string.split('').each_with_index do |char, idx|
    subs << char
    fragment = char

    string[idx + 1..-1].split('').each_with_index do |letter|
      fragment += letter
      subs << fragment
    end
  end

  subs
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: ['a', 'ab', 'abc', 'b', 'bc', 'c']"
  print "=> " 
  print consecutive_substrings('abc')

  puts

  puts "Expecting: ['a']"
  print "=> " 
  print consecutive_substrings('a')

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: javascript/consecutive_substrings.js
function consecutiveSubstrings(string) {
  // type your code here
  const stringArr = string.split('')
  const subs = []

  stringArr.forEach((char, idx) => {
    subs.push(char)
    let fragment = char

    stringArr.slice(idx + 1).forEach((letter) => {
      fragment = fragment + letter
      subs.push(fragment)
    })
  })

  return subs
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: ['a', 'ab', 'abc', 'b', 'bc', 'c']");
  console.log("=>", consecutiveSubstrings('abc'));

  console.log("");

  console.log("Expecting: ['a']");
  console.log("=>", consecutiveSubstrings('a'));
}

module.exports = consecutiveSubstrings;

// Please add your pseudocode to this file
// And a written explanation of your solution
